platform/linux-generic/pktio/null.c

```diff
@@ -0,0 +1,214 @@
+#include <net/if.h>
+#include <inttypes.h>
+#include <poll.h>
+#include <sys/ioctl.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <linux/ethtool.h>
+#include <linux/sockios.h>
+
+#include <odp_api.h>
+#include <odp_packet_socket.h>
+#include <odp_packet_internal.h>
+#include <odp_packet_io_internal.h>
+#include <odp_align_internal.h>
+#include <odp_debug_internal.h>
+#include <odp_classification_datamodel.h>
+#include <odp_classification_inlines.h>
+#include <odp_classification_internal.h>
+#include <odp/api/hints.h>
+
+#include <protocols/eth.h>
+#include <protocols/ip.h>
+
+static int null_close(pktio_entry_t *pktio_entry ODP_UNUSED)
+{
+	return 0;
+}
+
+static int null_open(odp_pktio_t id ODP_UNUSED,
+		     pktio_entry_t *pktio_entry ODP_UNUSED,
+		     const char *devname, odp_pool_t pool ODP_UNUSED)
+{
+	if (strncmp(devname, "null:", 5) != 0)
+		return -1;
+	return 0;
+}
+
+static int null_recv(pktio_entry_t *pktio_entry ODP_UNUSED,
+		     int index ODP_UNUSED, odp_packet_t pkt_table[] ODP_UNUSED,
+		     int len ODP_UNUSED)
+{
+	return 0;
+}
+
+static int null_fd_set(pktio_entry_t *pktio_entry ODP_UNUSED,
+		       int index ODP_UNUSED, fd_set *readfds)
+{
+	/* Somebody might wonder why readfds was not marked ODP_UNUSED.
+	 * The reason is separately run checkpatch script complains that
+	 * there must be spaces around '*', and if there are spaces, then
+	 * the continuous integration build checkpatch script complains
+	 * there must not be spaces around '*'. */
+	(void)readfds;
+	return 0;
+}
+
+static int null_recv_tmo(pktio_entry_t *pktio_entry ODP_UNUSED,
+			 int index ODP_UNUSED,
+			 odp_packet_t pkt_table[] ODP_UNUSED,
+			 int num ODP_UNUSED, uint64_t usecs)
+{
+	struct timeval timeout;
+	int maxfd = -1;
+	fd_set readfds;
+
+	timeout.tv_sec = usecs / 1000 / 1000;
+	timeout.tv_usec = usecs % (1000 * 1000);
+	FD_ZERO(&readfds);
+
+	select(maxfd + 1, &readfds, NULL, NULL,
+	       usecs == ODP_PKTIN_WAIT ? NULL : &timeout);
+
+	return 0;
+}
+
+static int null_recv_mq_tmo(pktio_entry_t *pktio_entry[] ODP_UNUSED,
+			    int index[] ODP_UNUSED, int num_q ODP_UNUSED,
+			    odp_packet_t pkt_table[] ODP_UNUSED,
+			    int num ODP_UNUSED, unsigned *from ODP_UNUSED,
+			    uint64_t usecs)
+{
+	struct timeval timeout;
+	int maxfd = -1;
+	fd_set readfds;
+
+	timeout.tv_sec = usecs / 1000 / 1000;
+	timeout.tv_usec = usecs % (1000 * 1000);
+
+	FD_ZERO(&readfds);
+
+	select(maxfd + 1, &readfds, NULL, NULL,
+	       usecs == ODP_PKTIN_WAIT ? NULL : &timeout);
+
+	return 0;
+}
+
+static int null_send(pktio_entry_t *pktio_entry ODP_UNUSED,
+		     int index ODP_UNUSED, const odp_packet_t pkt_table[],
+		     int len)
+{
+	int i;
+
+	for (i = 0; i < len; i++)
+		odp_packet_free(pkt_table[i]);
+
+	return len;
+}
+
+#define PKTIO_NULL_MTU (64 * 1024)
+
+static uint32_t null_mtu_get(pktio_entry_t *pktio_entry ODP_UNUSED)
+{
+	return PKTIO_NULL_MTU;
+}
+
+static const char null_mac[] = {0x02, 0xe9, 0x34, 0x80, 0x73, 0x05};
```

Comment:
Where does this "magic number" come from? Why not `00:00:00:00:00:00`?

Reply:
Maybe we should add internal API to generate random (or semi-random) local MAC-addresses with L bit being set?

Reply:
This is taken from the .pcap pktio, and incremented by one to make it distinct from the .pcap MAC address. But agree, an internal API to generate the MAC address would be great. It needs to be made sure the multicast bit is not being set, then.

Reply:
@lumag  any generation thing is bad when you need to reproduce something. I think all zeroes is good option here.

Reply:
@muvarov 
At least it should have local bit set on. I was thinking about using part of pktio name to generate mac.

Reply:
Need to decide what's needed here.

Reply:
From other side it's not clear why VMs do not have this address set https://www.techrepublic.com/blog/data-center/mac-address-scorecard-for-common-virtual-machine-platforms/
Docker has 02:42:ae:7f:27:e9

but linux lo does not have any mac at all. Maybe do null we also don't need any mac?

Reply:
The local bit is already set in the address. The binary representation shown is in inverse order, i.e. bit0, bit1, bit2, bit3, bit4, bit5, bit6, bit7 whereas the hexadecimal representation has bit7, bit6, bit5, bit4, bit3, bit2, bit1, bit0. So the local bit is (1<<1), i.e. 2. The 0x02 has the local bit and it also does not have the group bit, so it is a non-multicast locally administrated address.